doc8/main.py

```diff
@@ -256,79 +253,184 @@ def validate(cfg, files):
                                 "    - %s:%s: %s %s"
                                 % (f.filename, line_num, code, message)
                             )
-                        else:
+                        elif not result.capture:
                             print(
                                 "%s:%s: %s %s" % (f.filename, line_num, code, message)
                             )
+                        result.error(check_name, f.filename, line_num, code, message)
                         error_counts[check_name] += 1
             else:
                 raise TypeError("Unknown check type: %s, %s" % (type(c), c))
     return error_counts
 
 
+def get_defaults():
+    return {
+        "paths": [os.getcwd()],
+        "config": [],
+        "allow_long_titles": False,
+        "ignore": [],
+        "sphinx": True,
+        "ignore_path": [],
+        "ignore_path_errors": [],
+        "default_extension": "",
+        "file_encoding": "",
+        "max_line_length": MAX_LINE_LENGTH,
+        "extension": list(FILE_PATTERNS),
+        "quiet": False,
+        "verbose": False,
+        "version": False,
+    }
+
+
+class Result(object):
+    def __init__(self):
+        self.files_selected = 0
+        self.files_ignored = 0
+        self.error_counts = {}
+        self.errors = []
+        self.capture = False
+
+    @property
+    def total_errors(self):
+        return len(self.errors)
+
+    def error(self, check_name, filename, line_num, code, message):
+        self.errors.append((check_name, filename, line_num, code, message))
+
+    def finish(self, files_selected, files_ignored, error_counts):
+        self.files_selected = files_selected
+        self.files_ignored = files_ignored
+        self.error_counts = error_counts
+
+    def report(self):
+        lines = []
+        if self.capture:
+            for error in self.errors:
+                lines.append("%s:%s: %s %s" % error[1:])
+
+        lines.extend(
+            [
+                "=" * 8,
+                "Total files scanned = %s" % (self.files_selected),
+                "Total files ignored = %s" % (self.files_ignored),
+                "Total accumulated errors = %s" % (self.total_errors),
+            ]
+        )
+
+        if self.error_counts:
+            lines.append("Detailed error counts:")
+            for check_name in sorted(six.iterkeys(self.error_counts)):
+                check_errors = self.error_counts[check_name]
+                lines.append("    - %s = %s" % (check_name, check_errors))
+
+        return "\n".join(lines)
+
+
+def doc8(args=None, **kwargs):
```

Comment:
Personally, I'd really like to see this split out into separate kwargs with their own defaults. Something like:

    def doc8(paths=None, config=None, allow_long_titles=False, ...
             ...):
        if paths is None:
            paths = [os.getcwd()]
        config = config or []
        ...

I realize this would result in a bit more duplication but I think it makes things much clearer. Assuming you agree, could we do this as a follow-up?